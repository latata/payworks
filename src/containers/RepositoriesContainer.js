import { Container } from 'unstated';

import repositoriesService from '../services/repositories';

export default class RepositoriesContainer extends Container {
  state = {
    loadedOrganization: '',
    repositories: [],
    availableLanguages: [],
    language: '',
    organization: 'payworks',
    status: 'uninitialized'
  };

  async fetchData(org) {
    if (this.state.status === 'loaded' && org === this.state.loadedOrganization) {
      return;
    }

    this.setState({
      repositories: [],
      loadedOrganization: '',
      availableLanguages: [],
      language: '',
      status: 'loading',
    });

    try {
      const fetchedRepositories = await repositoriesService.fetchAll(org);
      let repositories = fetchedRepositories.sort((a, b) => b.stargazers_count - a.stargazers_count);

      const availableLanguages = fetchedRepositories.reduce((languages, { language }) => {
        if (language && languages.indexOf(language) === -1) {
          languages.push(language);
        }

        return languages;
      }, []).sort();

      this.setState({
        repositories,
        availableLanguages,
        loadedOrganization: org,
        status: 'loaded',
      });
    } catch (error) {
      if (error.response && error.response.status === 404) {
        this.setState({
          status: 'not-found'
        });
      } else {
        this.setState({
          status: 'error'
        });
      }
    }
  };
}
