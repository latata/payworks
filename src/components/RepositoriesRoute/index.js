import React, { useEffect } from 'react';
import connect from 'unstated-connect';
import Repositories from '../Repositories';
import RepositoriesContainer from '../../containers/RepositoriesContainer';
import useDebounce from '../../hooks/useDebounce';
import Spinner from '../Spinner';
import Filters from '../Filters';

function RepositoriesRoute({ containers: [repositoriesContainer] }) {
  const { state: { availableLanguages, language, organization, status } } = repositoriesContainer;
  let { repositories } = repositoriesContainer.state;
  // debounced organization to prevent XHR on each organization input change
  const [debouncedOrganization, setDebounceOrganization] = useDebounce(organization, 500);
  // empty string means all
  const allLanguages = ['', ...availableLanguages];

  const onOrganizationChange = (organization) => {
    repositoriesContainer.setState({ organization });
    setDebounceOrganization(organization);
  };

  const onLanguageChange = (language) => {
    repositoriesContainer.setState({ language });
  };

  // fetch repositories on component mount and on debouncedOrganization change
  useEffect(() => {
    repositoriesContainer.fetchData(debouncedOrganization);
  }, [debouncedOrganization, repositoriesContainer]);

  // filter by language
  if (language) {
    repositories = repositories.filter((repo) => {
      return repo.language === language;
    });
  }

  return (
    <div className="container">
      <h1>
        Organization's repositories
      </h1>
      <Filters
        organization={organization}
        language={language}
        languages={allLanguages}
        onLanguageChange={onLanguageChange}
        onOrganizationChange={onOrganizationChange}
      />
      {status === 'loading' && <Spinner />}
      {status === 'loaded' && <Repositories repositories={repositories} />}
      {status === 'error' && <div>Unexpected error</div>}
      {status === 'not-found' && <div>Organization does not exist</div>}
    </div>
  );
}

export default connect([RepositoriesContainer])(RepositoriesRoute);
