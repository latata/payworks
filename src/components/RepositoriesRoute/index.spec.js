import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { Provider } from 'unstated';
import debounce from 'lodash.debounce';
import axios from 'axios';
import AxiosMockAdapter from 'axios-mock-adapter';
import { MemoryRouter } from 'react-router-dom';

import RepositoriesRoute from './index';
import RepositoriesContainer from '../../containers/RepositoriesContainer';

// run debounced function immediately
jest.mock('lodash.debounce', () => jest.fn(fn => fn));

const axiosMock = new AxiosMockAdapter(axios);
const flushPromises = () => new Promise(resolve => setTimeout(resolve));

axiosMock.onAny().reply(200, [
  {
    name: 'repo',
    stargazers_count: 1,
    language: 'javascript',
  },
  {
    name: 'repo2',
    stargazers_count: 2,
    language: 'php',
  }
]);

function renderComponent(repositoriesContainer) {
  return render(
    <Provider inject={[repositoriesContainer]}>
      <MemoryRouter>
        <RepositoriesRoute />
      </MemoryRouter>
    </Provider>
  );
}

afterEach(cleanup);

test('RepositoriesRoute fetches repositories on mount and on organization update', async () => {
  const repositoriesContainer = new RepositoriesContainer();
  const fetchDataStub = jest.spyOn(repositoriesContainer, 'fetchData');
  const defaultOrg = 'payworks';
  const testOrg = 'test-org';

  const { getByLabelText, asFragment } = renderComponent(repositoriesContainer);

  expect(asFragment()).toMatchSnapshot();
  expect(repositoriesContainer.state.organization).toBe(defaultOrg);
  expect(fetchDataStub).toBeCalled();
  expect(fetchDataStub).toBeCalledWith(defaultOrg);

  fetchDataStub.mockReset();

  const orgInput = getByLabelText('Organization');

  await fireEvent.change(orgInput, {
    target: {
      value: testOrg,
    }
  });

  expect(repositoriesContainer.state.organization).toBe(testOrg);
  expect(debounce.mock.calls[0][1]).toBe(500);
  expect(fetchDataStub).toBeCalledWith(testOrg);

  // make sure mocked axios request replied
  await flushPromises();

  expect(asFragment()).toMatchSnapshot();
});

test('RepositoriesRoutes filters repositories by language', async () => {
  const repositoriesContainer = new RepositoriesContainer();
  const { getByLabelText, asFragment } = renderComponent(repositoriesContainer);

  // make sure mocked axios request replied
  await flushPromises();

  const languageSelect = getByLabelText('Language');

  await fireEvent.change(languageSelect, {
    target: {
      value: 'javascript',
    },
  });

  expect(repositoriesContainer.state.language).toBe('javascript');
  expect(asFragment()).toMatchSnapshot();
});

