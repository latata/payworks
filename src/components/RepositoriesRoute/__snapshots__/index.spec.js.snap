// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoriesRoute fetches repositories on mount and on organization update 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Organization's repositories
    </h1>
    <div
      class="row"
    >
      <div
        class="form-group col-sm"
      >
        <label
          for="organization"
        >
          Organization
        </label>
        <input
          class="form-control"
          id="organization"
          type="text"
          value="payworks"
        />
      </div>
      <div
        class="form-group col-sm"
      >
        <label
          for="language"
        >
          Language
        </label>
        <select
          class="form-control"
          id="language"
        >
          <option
            value=""
          >
            ALL
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RepositoriesRoute fetches repositories on mount and on organization update 2`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Organization's repositories
    </h1>
    <div
      class="row"
    >
      <div
        class="form-group col-sm"
      >
        <label
          for="organization"
        >
          Organization
        </label>
        <input
          class="form-control"
          id="organization"
          type="text"
          value="test-org"
        />
      </div>
      <div
        class="form-group col-sm"
      >
        <label
          for="language"
        >
          Language
        </label>
        <select
          class="form-control"
          id="language"
        >
          <option
            value=""
          >
            ALL
          </option>
          <option
            value="javascript"
          >
            javascript
          </option>
          <option
            value="php"
          >
            php
          </option>
        </select>
      </div>
    </div>
    <table
      class="table"
      data-testid="repositories-list"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Language
          </th>
          <th>
            Stars
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/repository/undefined"
            >
              repo2
            </a>
          </td>
          <td>
            php
          </td>
          <td>
            2
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/repository/undefined"
            >
              repo
            </a>
          </td>
          <td>
            javascript
          </td>
          <td>
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`RepositoriesRoutes filters repositories by language 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Organization's repositories
    </h1>
    <div
      class="row"
    >
      <div
        class="form-group col-sm"
      >
        <label
          for="organization"
        >
          Organization
        </label>
        <input
          class="form-control"
          id="organization"
          type="text"
          value="payworks"
        />
      </div>
      <div
        class="form-group col-sm"
      >
        <label
          for="language"
        >
          Language
        </label>
        <select
          class="form-control"
          id="language"
        >
          <option
            value=""
          >
            ALL
          </option>
          <option
            value="javascript"
          >
            javascript
          </option>
          <option
            value="php"
          >
            php
          </option>
        </select>
      </div>
    </div>
    <table
      class="table"
      data-testid="repositories-list"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Language
          </th>
          <th>
            Stars
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/repository/undefined"
            >
              repo
            </a>
          </td>
          <td>
            javascript
          </td>
          <td>
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
